core/src/main/java/org/apache/iceberg/BaseFilesTable.java

```diff
@@ -186,4 +201,43 @@ ManifestFile manifest() {
       return manifest;
     }
   }
+
+  static class ContentFileStructWithMetrics implements StructLike {
+    private final Schema projectedFilesSchema;
+    private final StructLike fileAsStruct;
+    private final Map<String, StructLike> readableMetrics;
+
+    ContentFileStructWithMetrics(
+        Schema projectedFileSchema,
+        StructLike fileAsStruct,
+        Map<String, StructLike> readableMetrics) {
+      this.projectedFilesSchema = projectedFileSchema;
+      this.fileAsStruct = fileAsStruct;
+      this.readableMetrics = readableMetrics;
+    }
+
+    @Override
+    public int size() {
+      return projectedFilesSchema.columns().size() + 1;
+    }
+
+    @Override
+    public <T> T get(int pos, Class<T> javaClass) {
+      if (pos < projectedFilesSchema.columns().size()) {
+        return fileAsStruct.get(pos, javaClass);
+      } else if (pos == projectedFilesSchema.columns().size()) {
+        return javaClass.cast(readableMetrics);
+      } else {
+        throw new IllegalArgumentException(
+            String.format(
+                "Illegal position access for FileRow: %d, max allowed is %d",
+                pos, fileAsStruct.size()));
+      }
+    }
+
+    @Override
+    public <T> void set(int pos, T value) {
+      throw new UnsupportedOperationException("FileEntryRow is read only");
```

Comment:
nit: What's `FileEntryRow`?

Reply:
Sorry, earlier version of name of this class